src/dict.c

```diff
@@ -124,6 +135,7 @@ uint64_t dictGenCaseHashFunction(const unsigned char *buf, size_t len) {
 #define ENTRY_PTR_MASK 7     /* 111 */
 #define ENTRY_PTR_NORMAL 0   /* 000 */
 #define ENTRY_PTR_NO_VALUE 2 /* 010 */
+#define ENTRY_PTR_EMBEDDED 4 /* 100 */
```

Comment:
[Readability suggestion]

looks like we are missing `1`. Is it currently used a key flag in `entryIsKey`? If so, it would be good to add it too.

Reply:
yes, `xx1` means "entry is key". Add a comment for it, or do you have another suggestion?

Reply:
maybe add a define? magic values are always hard to keep track of ...

```
#define ENTRY_PTR_IS_KEY 1
```

Reply:
Yeah, will add it. It's the trick @zuiderkwast used for entry is key as sds data pointer address is always an odd address because the metadata struct are all odd in size.

Reply:
The `ENTRY_PTR_NORMAL` and the others here are taken from the pointer's last three bits, i.e. it's masked with `ENTRY_PTR_MASK`.

The pointer-is-key flag is just one bit. What's the point of having a macro defined to 1? Do you also want to define a 1-bit mask for it too which explains which bits to use for this judgement?

    #define ENTRY_PTR_IS_KEY_MASK 1
    #define ENTRY_PTR_IS_KEY 1

I think that's excessive. These macros are only used in the next inline functions anyway. The inline functions make up the internal abstraction for these encoded pointers.

```C
static inline int entryIsKey(const dictEntry *de) {
    return (uintptr_t)(void *)de & 1;
}

static inline int entryIsNormal(const dictEntry *de) {
    return ((uintptr_t)(void *)de & ENTRY_PTR_MASK) == ENTRY_PTR_NORMAL;
}
```

@PingXie if you want an opaque abstraction of this, then I think the only way is to put this part in a separate file and only expose certain functions to it. Is it worth it?

Reply:
I'd say we shouldn't change this abstraction just because we're using one of the unused values here. What we could do is to add a comment like this, just to show that any pattern with the last bit = 1 are taken.

```suggestion
#define ENTRY_PTR_EMBEDDED 4 /* 100 */
/*      ENTRY_PTR_IS_KEY        xx1 */
```